### ISSUE TITLE
Pub/Sub: subscriber waiting on future.result() blocks forever

### ISSUE THREAD

**Issue by smemsh** [GoogleCloudPlatform/google-cloud-python#4621]
when pubsub endpoints changed recently, it broke our code (after working fine for about a year).  we have ported our code to the new async API in 0.29.2 to get it working again.

following the documentation, we're calling `.open()` to register a callback; when run, it stores the message locally, then calls `message.ack()`.  in the main loop, we're waiting on the Future's `.result()` to be returned so we can do some bottom-half processing on the stored message.  however, our subscriber blocks forever waiting for the Future's .result() method to return, and I don't think this is how it's supposed to work based on the documentation:

> The `pubsub_v1.subscriber.policy.thread.Policy.open` method returns a `pubsub_v1.subscriber.futures.Future`, which is both the interface to wait on messages (e.g. block the primary thread) and to address exceptions.
> 
> To block the thread you are in while messages are coming in the stream, use the `.pubsub_v1.subscriber.futures.Future.result` method [...] You can also use this for error handling; any exceptions that crop up on a thread will be set on the future.

if it's "the interface to wait on messages," we're quite puzzled, because it never returns any results; it seems to block forever.  note that the same call in the publisher *does not* block; it returns the published message ID right away, which is what we expect.  only the subscriber has this strange behavior of never returning the result.

distilled to simple test case, here's the publisher:
```python
topic_path = 'projects/ourproject/topics/test12345'
publisher = pubsub.PublisherClient()
publisher.publish(topic_path, b'New Message {0}'.format(datetime.now()))
```

and the subscriber (run this first to create the topic):
```python
def callback(message):
    print 'Callback triggered'
    print message
    message.ack()

topic_path = 'projects/ourproject/topics/test12345'
subscription_path = 'projects/ourproject/subscriptions/test12345'

publisher = pubsub.PublisherClient()
publisher.create_topic(topic_path)

subscriber = pubsub.SubscriberClient()
subscriber.create_subscription(subscription_path, topic_path)
subscription = subscriber.subscribe(subscription_path)

future = subscription.open(callback)
result = future.result()
```
^^^ this is where it hangs forever, although it continues to print messages when the callback runs, on each receive.  originally we thought it was because we're using a ThreadPool and calling these via `.map()`, it might be deadlocked somehow, but after reduction to a toy demo with single thread, single subscriber, no other work being done, it's still blocking forever.  here's the google lib versions:

```
 $ pip freeze | grep google
google-api-core==0.1.2
google-auth==1.2.1
google-auth-oauthlib==0.2.0
google-cloud-core==0.28.0
google-cloud-pubsub==0.29.2
google-compute-engine==2.7.1
googleapis-common-protos==1.5.3
grpc-google-iam-v1==0.11.4
```

are we using the interface incorrectly? is `future.result()` supposed to block forever? if we're supposed to do *all* the work in the subscriber callback, and just use `result` to catch exceptions later, then how do we actually exit the blocking call when all the work is done? we have noticed that doing our own `.set_result()` in the handler does cause `.result()` to return, are we supposed to be setting the result ourselves like this?

not sure if this is user error, or a bug.  I see the patch in #3888 that seems to be the one adding this feature of returning the Future, and the following quote:

> I do think it is fine to have a future that never has a result (or only gets one if you explicitly close)

so I'm guessing that we're using it wrong and supposed to only use this for exceptions.  If so, suggest some language might be added to the docs, that the Future should *not* be used to wait on messages themselves, but actually to wait on closure of the subscription.

**Comment by dhermes** [comment]
@smemsh Thanks for taking the time to file such a detailed report!

> are we using the interface incorrectly?

Yes, but I think this issue illustrates the need for better docstrings and probably a narrative use case.

The call

```python
future = subscription.open(callback)
```

has just opened a worker (actually a few workers) that will listen for messages on the `subscription` (which is a [`Policy`][3] instance). Once `open()`-ed, the `Policy` can stay open indefinitely. The **only** time that `future.result()` would be set is if the policy [gets `close()`-ed][4].

It's certainly expected that you'd call `subscription.close()` before shutting down your code that subscribes to a given topic.

So, at the end of it, the only **real** value for the future is in checking `future.done()` to make sure the workers are still running (if `done()` returns `False`) and also `future.exception()` to see if there was a failure that caused it to stop running.

Typical usage would be as in the [Pull Subscriber Guide][5]:

```python
policy = subscriber.subscribe(subscription_path)
future = policy.open(callback)

# The subscriber is non-blocking, so we must keep the main thread from
# exiting to allow it to process messages in the background.
print('Listening for messages on {}'.format(subscription_path))
while True:
    time.sleep(60)
```

i.e. your subscriber would just stay alive "forever" (hence the `time.sleep(60)`).

----

ALSO: Why are you not on `0.29.2` (there continue to be crucial bugfixes, see [`0.29.3` release notes][1] and [`0.29.4` release notes][2])?

ALSO: I notice your `print`-s will "break" completely in Python 3. Any reason you're still on Python 2?

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/releases/tag/pubsub-0.29.3
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/releases/tag/pubsub-0.29.4
[3]: https://googlecloudplatform.github.io/google-cloud-python/latest/pubsub/subscriber/api/policy.html#google.cloud.pubsub_v1.subscriber.policy.thread.Policy
[4]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/pubsub-0.29.4/pubsub/google/cloud/pubsub_v1/subscriber/policy/thread.py#L162-L164
[5]: https://cloud.google.com/pubsub/docs/pull#pubsub-pull-messages-async-python

**Comment by smemsh** [comment]
thanks for the info.  having to do all work in the callback will require restructuring our program somewhat, but should be ok, there's a particular kind of message that signals completion for the overall process, and then we want to do things which the callback receiving messages doesn't know how to do; we will adapt our usage.

may I ask why the publisher returns the result when it sends a message (it returns the message-id)? I didn't call any close method on that; are the publish events seen as individual in this case and the whole thing is done in one go, whereas subscription is a continuous event stream?

Re: documentation, I think the main thing that confused me from thinking the callbacks really did have to do all the work (and couldn't just do some of the work) was the specific:

> The open() method returns a Future, which is both the interface to wait on messages (e.g. block the primary thread) and to address exceptions.

I confused "messages" to mean "each of the messages" not "all of the messages," like looping forever getting more of them, each one blocking until another is available.  I thought the callback was just something offered/possible/available to do something right at reception (like stash it for later processing by a different thread), but not requiring the user to do all his work there.

0.29.4 already, the pace of change is quick, that's the version we pinned when we started making it work with the async API, had to be less than two weeks ago.  hope it's a keeper for the golden image that we won't change for another few years ;-) half kidding...

python2 print statements, have less chars in toy example.  I'll reconsider in future posts, thanks.

**Comment by dhermes** [comment]
If you want your callback to know about the policy it is "working for", I'd suggest something like:

```python
class Callback(object):

    def __init__(self):
        self.close_event = threading.Event()

    @property
    def done(self):
        return self.close_event.is_set()

    def __call__(self, message):
        if is_done(message):
            self.close_event.set()
        return handle_message(message)


def main():
    policy = subscriber.subscription(...)
    callback = Callback()
    future = policy.open(callback)

    while True:
        time.sleep(15)
        if callback.done:
            break

    policy.close()
```

----

> may I ask why the publisher returns the result when it sends a message

I assume you mean the following:

```python
future = publisher.publish(topic_path, b'some data', moo='lah')
message_id = future.result()
```

The "best" answer might be that those futures (i.e. the ones from `PublisherClient.publish()`) actually have something to return (we [do a "sanity check"][1] that each message in the publish corresponds to a future on the batch).

>  I didn't call any close method on that

By default, when you call `publisher.publish(...)` it actually pushes the message onto a `Batch` (rather than blocking the main thread). That batch has a short lifespan (0.05s by default) and once that lifespan is over, the batch will commit in a worker thread (that was spawned and sleeping since the batch was created).

> are the publish events seen as individual in this case and the whole thing is done in one go, whereas subscription is a continuous event stream?

A `publish()` is something "bounded", i.e. here is exactly one message, please publish it. Whereas a subscription is an open-ended stream of messages from a given topic, both the number of messages and the lifetime of the topic are unknown / unbounded.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/pubsub-0.29.4/pubsub/google/cloud/pubsub_v1/publisher/batch/thread.py#L194

**Comment by smemsh** [comment]
I have moved our work into the callback, and upgraded to 0.30.0, and it seems to work.  however, at the end of processing, subscriber `.close()` still does not unblock `future.result()`, as I think it's supposed to based on your earlier comment:

> The *only* time that `future.result()` would be set is if the policy gets close()-ed.

I'm calling `.close()` in the subscriber callback when we see a completion message from our publisher, which should cause the main thread (blocked on `.result()`) to return, but it remains hung indefinitely.

can you clarify that waiting on `.result()` should unblock upon subscription.close() ?

**Comment by dhermes** [comment]
Yes I can confirm. See the [source][1]. A few questions

1. Are you saying that `.close()` actually exits?
1. Where are you calling `.close()`? You may have a thread chasing it's own tail. (There is a reason I recommended setting a `close_event` in your callback, since the policy can't be closed from within it's own worker threads.
1. (It seems like this is answered, but I'll ask anyway.) Are you saying it blocks or just that `.result() is None`?

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/pubsub-0.30.0/pubsub/google/cloud/pubsub_v1/subscriber/policy/thread.py#L181-L182

**Comment by smemsh** [comment]
indeed, was running `.close()` from callback, which would have been run by the subscriber policy thread.  I determined that `.close()` was in fact not returning as you mentioned it might not

following your earlier suggestion I implemented the thread.Event(), using an `is_set()` in main loop to check if we should run the `.close()` method.  otherwise I loop on `future.result(timeout=1)` checking for `future.done()`.  keeping with waits on `future.result()` (but adding timeout) instead of just sleep loop lets the bubbled up exception get raised, while still letting me look for the done event set by callback.

everything is working now, we are good.  thanks for walking us through it.
